src/subtype.c

```diff
@@ -1301,24 +1483,37 @@ JL_DLLEXPORT int jl_obvious_subtype(jl_value_t *x, jl_value_t *y, int *subtype)
                     return 1;
             }
             int i, npx = jl_nparams(x), npy = jl_nparams(y);
-            int vx = 0, vy = 0;
+            jl_vararg_kind_t vx = JL_VARARG_NONE;
+            jl_value_t *vxt = NULL;
+            int vy = 0;
+            int vnpx = npx;
             if (istuple) {
-                vx = npx > 0 && jl_is_vararg_type(jl_tparam(x, npx - 1));
+                if (npx > 0) {
+                    jl_value_t *xva = jl_tparam(x, npx - 1);
+                    vx = jl_vararg_kind(xva);
+                    if (vx != JL_VARARG_NONE) {
+                        vxt = jl_unwrap_vararg(xva);
+                        vnpx -= 1;
+                        if (vx == JL_VARARG_INT)
+                            vnpx += jl_vararg_length(xva);
+                    }
+                }
                 vy = npy > 0 && jl_is_vararg_type(jl_tparam(y, npy - 1));
             }
-            if (npx != npy || vx || vy) {
-                if (!vy) {
+            if (npx != npy || vx != JL_VARARG_NONE || vy) {
+                if ((vx == JL_VARARG_NONE || vx == JL_VARARG_UNBOUND) && !vy) {
```

Comment:
I think `UNBOUND` is computed incorrectly for subtyping's use (it was introduced for gf.c's heuristic type signature limiting), since it doesn't count number of uses. Consider `Tuple{Vararg{NTuple{N}, N} where N}`. To be conservative, just consider `jl_is_long` vs `jl_is_typevar` in this fast-path.

Reply:
What the problem with that case? If the RHS is not a vararg, then if we have a unionall here it can't possibly be a subtype.

Reply:
I don't know. I just don't trust this code to do any counting of TypeVar uses.